src/vm.c

```diff
@@ -213,13 +215,10 @@ int vm_run(vm_t *v)
 
         switch (run->exit_reason) {
         case KVM_EXIT_IO:
-            if (run->io.port == 0x3f8 && run->io.direction == KVM_EXIT_IO_OUT) {
-                uint32_t size = run->io.size;
-                uint64_t offset = run->io.data_offset;
-                printf("%.*s", size * run->io.count, (char *) run + offset);
-            } else if ((run->io.port == 0x3f8 + 5) &&
-                       (run->io.direction == KVM_EXIT_IO_IN))
-                *((char *) run + run->io.data_offset) = 0x20;
+            if (run->io.port >= COM1_PORT_BASE && run->io.port < COM1_PORT_END)
+                serial_handle(&v->serial, run->io.direction, run->io.port,
```

Comment:
If function prototype of `serial_handle` is changed to two parameters, we can avoid the data copy overhead. That is,
```
void serial_handle(serial_dev_t *s, struct kvm_run *r);
```

Eventually, the object of `struct kvm_run` will be updated including its internal data.